### PR TITLE
Improvements for handling string parameters

### PR DESCRIPTION
**Pull request by authaldo**
In the current version of rig_reconfigure, every character modification within a string parameter causes an update which is sent to the respective node.

This MR modifies the behaviour:
An update is now only sent to the node if the editing is finished by pressing enter or the input field is deselected. For the future it might also be interesting to include a timeout which handles situations where the user has finished editing but forgets to press enter (and doesn't change the focus within the window). 
The timeout is not part of this MR but might be added in the future. Given the low usage of string parameters within our internal software stack, the timeout is currently solely a low priority enhancement.

The MR further includes minor layout changes which ensure that nesting doesn't cause textfields to span across the border of the window (if the window is large enough to handle the top level text fields).

### CODE REVIEW

File: src/rig_reconfigure.cpp [hunk 1]
```diff
@@ -22,13 +22,21 @@
...
+/// Reduction of the text field width per nesting level (necessary to avoid input field spanning across the window
+/// borders)
+constexpr auto TEXT_FIELD_WIDTH_REDUCTION_PER_LEVEL = 22;
 constexpr auto FILTER_INPUT_TEXT_FIELD_WIDTH = 250;
 constexpr auto FILTER_HIGHLIGHTING_COLOR = ImVec4(1, 0, 0, 1);
 constexpr auto STATUS_WARNING_COLOR = ImVec4(1, 0, 0, 1);
 constexpr auto NODES_AUTO_REFRESH_INTERVAL = 1s; // unit: seconds
 constexpr auto DESIRED_FRAME_RATE = 30;
 constexpr std::chrono::duration<float> DESIRED_FRAME_DURATION_MS = 1000ms / DESIRED_FRAME_RATE;
+constexpr auto TEXT_INPUT_EDITING_END_CHARACTERS = "\n";
```
Comment (ottojo) [inline src/rig_reconfigure.cpp:39]:
```suggestion
constexpr char TEXT_INPUT_EDITING_END_CHARACTERS = '\n';
```

File: src/rig_reconfigure.cpp [hunk 2]
```diff
@@ -486,7 +503,25 @@ std::set<ImGuiID> visualizeParameters(ServiceWrapper &serviceWrapper,
                         std::make_shared<ParameterModificationRequest>(ROSParameter(fullPath, value)));
             }
         } else if (std::holds_alternative<std::string>(value)) {
+            bool flush = false;
```
Comment (ottojo) [inline src/rig_reconfigure.cpp:506]:
```suggestion
            // Set to true when enter is pressed
            bool flush = false;
```

File: src/rig_reconfigure.cpp [hunk 3]
```diff
@@ -486,7 +503,25 @@ std::set<ImGuiID> visualizeParameters(ServiceWrapper &serviceWrapper,
...
+                auto &str = std::get<std::string>(value);
+
+                // check if last character indicates the end of the editing
+                if (str.ends_with(TEXT_INPUT_EDITING_END_CHARACTERS)) {
+                    flush = true;
+                    str.pop_back();
+                }
+            }
+
+            if (flush || (!ImGui::IsItemActive() && dirtyTextInput == fullPath)) {
```
Comment (ottojo) [inline src/rig_reconfigure.cpp:523]:
```suggestion
            // Second condition: InputText focus lost
            if (flush || (!ImGui::IsItemActive() && dirtyTextInput == fullPath)) {
```

File: src/rig_reconfigure.cpp (hunk 1 above)
Comment (authaldo) [reply src/rig_reconfigure.cpp:39]: Das war eigentlich schon absichtlich ein string für den Fall, dass man bspw. auf Windows auch auf '\r' achten muss (habe ich aber nicht vor zu testen :D)

Comment (ottojo) [reply src/rig_reconfigure.cpp:39]: alright, ist sicher überflüssig aber whatever, verschwendet sicher nicht messbar viele cpu cycles

**Review by ottojo — APPROVED** [review]

**Pull request merged**